enhancements/network/persistent-ips-sdn-secondary-networks.md

```diff
@@ -0,0 +1,743 @@
+also need to ensure the pod where the VM runs was deleted. If it weren't, we
+will end up in a situation where the IP would be returned to the pool earlier
+than it should, which would lead to the same IP in two workloads at the same
+time: one of them shutting down, the other already readily processing traffic.
+
+The garbage collection mechanism will behave differently depending on the
+[DeletionPropagation](https://github.com/kubernetes/apimachinery/blob/3e2600dc79feea6cdc8a9224bc8a6a7fcfee1466/pkg/apis/meta/v1/types.go#L526)
+configured by the user on the `DeleteOptions` provided when deleting the VM.
+
+#### Deleting the VM with the Background `DeletionPropagation`
+
+The following sequence diagram describes how Kubernetes GC works for the
+KubeVirt entities (VM / VMI / pod):
+
+```mermaid
+sequenceDiagram
+  actor User
+  participant persistentIP
+  participant GC
+  participant VM
+  participant VMI
+  participant launcherPod
+
+  User->>VM: Delete VM
+  VM->>VMI: Delete VMI
+  activate VM
+  VMI->>launcherPod: Delete launcherPod
+  activate VMI
+  launcherPod-->>VMI: launcherPod Deleted
+  Deactivate VMI
+  note over VMI: clear VMI finalizer
+  VMI-->>VM: VMI Deleted
+  Deactivate VM
+  note over VM: clear VM finalizer
+  VM-->>User: VM Deleted
+  VM->>GC: VM deleted
+  GC->>persistentIP: Delete persistentIP
+  persistentIP-->>GC: persistentIP Deleted
+```
+
+The VM has a finalizer that prevents it from being deleted while the Virtual
+Machine Instance (VMI, represents a running VM) is around; the VMI - in turn -
+has a finalizer that prevents it from being deleted until the pod where the VM
+runs is deleted. These two ensure the VM object is only removed from the
+datastore after the VM has actually stopped. Thus, in this scenario, we have
+avoided the potential issue described above.
+
+#### Deleting the VM with the Foreground `DeletionPropagation`
+
+When using the foreground `DeletionPropagation`, the dependent object's - e.g.
+the VMI, and the persistentIP allocations - deletion will be triggered
+**before** the VM is actually deleted.
+
+Given the existing finalizers set in the VM and VMI object, the VM object will
+not be deleted while the pod is in the datastore - but the corresponding
+`PersistentIP` CR will be scheduled for deletion.
+
+Thus, when the foreground deletion propagation is chosen, the IP is returned to
+the pool while the VM is still shutting down (or about to start to ...), which
+will lead to the probable scenario described above: another workload will be
+assigned the IP for a VM which is still running.
+
+#### Deleting the VM with the Orphaned `DeletionPropagation`
+
+When using the orphaned `DeletionPropagation`, the dependent objects - e.g.
+will **not** be deleted by the garbage collector. I.e. the user is asking for
+the dependent resources to leak.
+
+#### Recommendation
+
+The simplest way to work-around this issue is simply to only allow `Background`
+propagation deletions.
```

Comment:
Worth to mention that kubevirt delete behaviour always does "foreground" behind the scene (unless orphan is used)
Because of 2 things:
1. The finalizers that kubevirt manages - they will enforce that resources are deleted bottom up (pod vmi vm).
2. Due to [1] kubevirt also must propogate the delete signal itself, i.e VM will trigger VMI delete when VM is deleted, and VMI trigger pod delete, else the finalizers would never be removed.

It means that we actually don't lose any feature by blocking foreground.

Reply:
> It means that we actually don't lose any feature by blocking foreground.

I think this would matter only to KubeVirt maintainers. Thus this info, if needed, should be on a KubeVirt design proposal, rather than here.

Reply:
I am again thinking about this ...

Who actually invokes the VMI / pod deletes ? Isn't it the VM / VMI controller ? Thus, it means the object is already removed from the data-store.

For it to be like the behavior you describe, the delete code would have to be on a webhook IIUC.

Reply:
The object isn't removed from the data-store, it has finalizers, it can't be removed until they are removed
here the VM controller that does all this
https://github.com/kubevirt/kubevirt/blob/4accdebe16489462937e2d52f8e233736a8acfa4/pkg/virt-controller/watch/vm.go#L2593

Reply:
Right, thanks for keeping me honest.